static/graph-layout-core.ts

```diff
@@ -330,44 +371,53 @@ export class GraphLayoutCore {
             block.row = 0;
             block.col = 0;
             block.boundingBox = {
-                rows: 1,
-                cols: 2,
+                width: 2,
+                height: 1,
+                rows: [{start: 0, end: 2}],
             };
         } else if (block.treeEdges.length === 1) {
             const childIndex = block.treeEdges[0];
             const child = this.blocks[childIndex];
             block.row = 0;
             block.col = child.col;
             block.boundingBox = {
-                rows: 1 + child.boundingBox.rows,
-                cols: child.boundingBox.cols,
+                width: child.boundingBox.width,
+                height: child.boundingBox.height + 1,
+                rows: [
+                    {start: child.col, end: child.col + 2},
+                    ...child.boundingBox.rows.map(bound => cloneDeep(bound)),
+                ],
```

Comment:
There's some `O(n^2)` triangularness going on here, but, everything about the tree updates are `O(n^2)`. The main way to make it not `O(n^2)`, or at least less so, would be to clear the information from the children since the information for full subtree is all that's needed.